runtime/android/core_internal/src/org/xwalk/core/internal/ReflectionHelper.java

```diff
@@ -117,6 +154,145 @@ static boolean shouldUseLibrary() {
         }
         return false;
     }
+
+    // We are doing following checks before loading shared library:
+    //  1. The APK contains only one Certificate, this is to protect from FakeID attack.
+    //  2. The contained certificate is exactly the one expected to be.
+    //  3. additional check for master key exploit since Android 4.0.
+    //     From http://androidvulnerabilities.org/by/manufacturer/all,
+    //     There are three master key exploit since Android 4.0.
+    public static boolean checkLibrary(Context app, Context library) throws NameNotFoundException, IOException {
+        String apk = library.getPackageCodePath();
+        if (apk == null || apk.isEmpty()) return false;
+        JarFile apkFile = new JarFile(apk);
+        Enumeration<JarEntry> entries = apkFile.entries();
+        int rsaCount = 0;
+        boolean rsaMatched = false;
+        boolean hasDuplicateEntries = false;
+        Set<String> entryNames = new HashSet<String>();
+        while(entries.hasMoreElements()) {
+            JarEntry entry = entries.nextElement();
+            String entryName = entry.getName();
+            if (entryNames.contains(entryName)) {
+                hasDuplicateEntries = true;
+                break;
+            } else {
+                entryNames.add(entryName);
+            }
+            if (entryName.toUpperCase().startsWith("META-INF/") && entryName.toUpperCase().endsWith(".RSA")) {
+                rsaCount ++;
+                InputStream is = apkFile.getInputStream(entry);
+                byte[] tmpBuffer = new byte[1024];
+                ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+                int count;
+                while ((count = is.read(tmpBuffer, 0, 1024)) != -1) {
+                    buffer.write(tmpBuffer, 0, count);
+                }
+                buffer.flush();
+                is.close();
+                Signature sig = new Signature(buffer.toByteArray());
+                String sigContent = sig.toCharsString();
+                // TODO(wang16): Compare with expected RSA
+                if (sigContent.isEmpty()) continue;
+                rsaMatched = true;
+            }
+        }
+        apkFile.close();
+        // Crosswalk Library apk should only contain one certificate.
+        // Multiple signature introduces fake id risk.
+        if (rsaCount != 1) return false;
+        // For master key exploit 8219321, it's using multiple entries with same name to bypass system checking.
+        // If there is no duplicate entries, we are safe.
+        if (hasDuplicateEntries) return false;
```

Comment:
The duplicates entries are not only limited to RSA files. And the check here for duplicated entries are not against RSA file either. It's prevent duplicate `classes.dex`, `libxwalkcore.so` and etc.

For legit multiple signatures, we can extends the implementation to multiple, just that we must know exactly how many RSAs will be there and the content of each of them.